### PR TITLE
Allow disabling custom passes.

### PR DESCRIPTION
**Pull request by csarofeen**
Right now there's no way to remove a custom pass once it has been added. This PR allows that to happen, and adds an abstract class that makes it easy to enable for passes.

### CODE REVIEW

**Comment by csarofeen** [comment]
Included the renaming from https://github.com/pytorch/pytorch/pull/33674

**Review by ZolotukhinM — COMMENTED** [review]
Thanks for adding this, I think it's a really useful feature to have! Could you please elaborate more on the implementation details? I feel it's a little hard to follow at them moment (or maybe harder than it could be). Maybe it is possible to simplify it a bit, or at least add some comments explaining why it is done this way?

File: torch/csrc/jit/passes/pass_manager.h [hunk 1]
```diff
@@ -22,18 +22,56 @@ namespace jit {
...
 
-struct TORCH_API RegisterPreFusionPass {
-  RegisterPreFusionPass(GraphPass p);
+struct TORCH_API RegisterPrePass {
+  // Back-compat
+  RegisterPrePass(GraphPass p);
+  static GraphPassNameType registerPrePass(GraphPass p);
+};
+
+struct TORCH_API ClearPostPass {
```
[the remark this answers is not in the document]
Comment (csarofeen) [reply torch/csrc/jit/passes/pass_manager.h]: I don't believe it does, was keeping it consistent with how it was. I wasn't wasn't sure if it was a struct for any particular reason before.

```
struct TORCH_API RegisterPreFusionPass {
  RegisterPreFusionPass(GraphPass p);
}
```
is where we started from. I couldn't see a technical reason for it immediately but assumed there was one. If you'd like I can try to remove the structs.

File: torch/csrc/jit/passes/pass_manager.h [hunk 2]
```diff
@@ -23,17 +23,77 @@ namespace jit {
...
+   * have to worry about initializing the static members.
+   */
+  static bool isRegistered(bool flip_bit = false);
+  /*
+   * name() will return the name of the registered pass
+   * name(pass_name, true) will set the name of the pass
+   * Similarly to isRegistered we use an internal static variable to hold the
+   * name.
+   */
+  static GraphPassNameType name(
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/pass_manager.h]: Similarly, can we make `name` a static member of the class? Also, a bit of bikeshedding: should it be called `pass_id` (when I first read the code and saw "name" I expected it to be a string)?

File: torch/csrc/jit/passes/pass_manager.h [hunk 3]
```diff
@@ -23,17 +23,77 @@ namespace jit {
...
+ protected:
+  /*
+   * isRegistered() will return if a pass has been registered
+   * isRegistered(true) will change the value of the internal static bool
+   *
+   * There's an internal static bool to this function to keep track of the
+   * state, this is so when functions are derived from this class, they don't
+   * have to worry about initializing the static members.
+   */
+  static bool isRegistered(bool flip_bit = false);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/pass_manager.h]: Can we make `isRegistered` a static member of this class and add getters/setters for it instead of one function that does both?

File: torch/csrc/jit/passes/pass_manager.h [hunk 4]
```diff
@@ -23,17 +23,77 @@ namespace jit {
...
+using GraphPassEntry = std::pair<GraphPass, GraphPassNameType>;
+
+// Return currently registered passes. Passes are stored in a static vector
+TORCH_API std::vector<std::pair<GraphPass, GraphPassNameType>>& getCustomPostPasses();
+TORCH_API std::vector<std::pair<GraphPass, GraphPassNameType>>& getCustomPrePasses();
+
+TORCH_API GraphPassNameType registerPostPass(GraphPass p);
+TORCH_API GraphPassNameType registerPrePass(GraphPass p);
+//Default to registerPostPass
+TORCH_API GraphPassNameType registerPass(GraphPass p);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/pass_manager.h]: I kept `registerPass` when added `registerPostPass`/`registerPrePass` to not break it for users who might have been using the old API. This PR changes its signature anyway, so we can remove `registerPass` altogether, no reason to keep it.

File: torch/csrc/jit/python/init.cpp [hunk 5]
```diff
@@ -360,7 +361,7 @@ void initJITBindings(PyObject* module) {
             checkAliasAnnotation(g, std::move(stack), unqualified_op_name);
           })
       .def(
-          "_jit_register_cuda_fuser", &registerCudaFuseGraph)
+          "_jit_register_cuda_fuser", &RegisterCudaFuseGraph::registerPass)
```
Comment (ZolotukhinM) [inline torch/csrc/jit/python/init.cpp]: You might want to expose `RegisterCudaFuseGraph::clearPass` as well.

File: torch/csrc/jit/passes/pass_manager.h [hunk 6]
```diff
@@ -23,17 +23,77 @@ namespace jit {
...
+  static GraphPassNameType name(
+      GraphPassNameType PassName = 0,
+      bool set = false);
 
-struct TORCH_API RegisterPreFusionPass {
-  RegisterPreFusionPass(GraphPass p);
+ public:
+  // registerPass(pass) will register the pass provided and set the
+  // name/isRegistered functions appropriately
+  static void registerPass(GraphPass p);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/pass_manager.h]: Could we allow registering both pre and post passes here?

**Review by ZolotukhinM — COMMENTED** [review]
Mostly looks good, please find some comments inline.

File: torch/csrc/jit/passes/pass_manager.h (hunk 2 above)
Comment (csarofeen) [reply torch/csrc/jit/passes/pass_manager.h]: name (now pass id) and isRegistered could be a member of the class, but that would require those deriving from the class to instantiate them in their .cpp file. There's two options here, either have additional small complexity in everyone using the pass manager (defining `MyPassRegister::pass_id = 0` and `MyPassRegister::is_registered = false`, or have them as static  vars in a function. I chose this way as now to use this register function you only have to modify your header file in a pretty concise manner.

I do prefer this way, but if you have a strong preference in having the instantiations in the .cpp and static member vars, I can do that.

File: torch/csrc/jit/passes/pass_manager.h (hunk 4 above)
Comment (csarofeen) [reply torch/csrc/jit/passes/pass_manager.h]: Removed.

File: torch/csrc/jit/python/init.cpp (hunk 5 above)
Comment (csarofeen) [reply torch/csrc/jit/python/init.cpp]: I need to fix merge conflicts in the big fuser PR. I'll do this then, so I can also test it. Let me know if you prefer it here.

File: torch/csrc/jit/passes/pass_manager.h (hunk 6 above)
Comment (csarofeen) [reply torch/csrc/jit/passes/pass_manager.h]: I couldn't think of an easy way to do this. One option could be another bool template parameter  that marks post or pre. Any thoughts?

**Comment by csarofeen** [comment]
Failures look unrelated to the pr

**Comment by csarofeen** [comment]
@peterjc123 Would you be able to look at the windows build? I can't figure out why it's not finding any of the symbols and @soumith thought you might be able to help.
